pkg/file/file_test.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package file_test
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/file/splitter"
+	test "github.com/ethersphere/bee/pkg/file/testing"
+	"github.com/ethersphere/bee/pkg/storage/mock"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+var (
+	start = 0
+	end   = test.GetVectorCount()
+)
+
+// TestSplitThenJoin splits a file with the splitter implementation and
+// joins it again with the joiner implementation, verifying that the
+// rebuilt data matches the original data that was split.
+//
+// It uses the same test vectors as the splitter tests to generate the
+// necessary data.
+func TestSplitThenJoin(t *testing.T) {
+	for i := start; i < end; i++ {
+		dataLengthStr := strconv.Itoa(i)
+		t.Run(dataLengthStr, testSplitThenJoin)
+	}
+}
+
+func testSplitThenJoin(t *testing.T) {
+	var (
+		paramstring = strings.Split(t.Name(), "/")
+		dataIdx, _  = strconv.ParseInt(paramstring[1], 10, 0)
+		store       = mock.NewStorer()
+		s           = splitter.NewSimpleSplitter(store)
+		j           = joiner.NewSimpleJoiner(store)
+		data, _     = test.GetVector(t, int(dataIdx))
+	)
+
+	// first split
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	dataReader := file.NewSimpleReadCloser(data)
+	resultAddress, err := s.Split(ctx, dataReader, int64(len(data)))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// then join
+	r, l, err := j.Join(ctx, resultAddress)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if l != int64(len(data)) {
+		t.Fatalf("data length return expected %d, got %d", len(data), l)
+	}
+
+	// read from joiner
+	var resultData []byte
+	for i := 0; i < len(data); i += swarm.ChunkSize {
```

Comment:
`ioutil.ReadAll()`?

Reply:
Sorry, no; joiner needs chunk-wise reads.

Reply:
hmmm. it would be nice if `Read` is then able to support this somehow. did you look at `ReadAll()` and why would it not work?

Reply:
Also please consider; if this is a 100GB file, for example, would you want to ReadAll it at all?

Reply:
Not sure if I understand this.

What would the end-user interface to the reads be, in a non-test scenario?

Reply:
Actually, with the amount of data in this test, it is not needed to have readers. But, once we have tests that are validating and measuring performance on large files, we may want to use readers approach. It is fine to keep these data slices in memory.